### DISCUSSION TITLE
Discord bot!

### DISCUSSION THREAD

**Discussion by thboss** [PhlexPlexico/G5API#80] [Show and tell]
80% completed, It just need to handle some exceptions
https://user-images.githubusercontent.com/40491968/110239013-d22c5300-7f55-11eb-9cdc-0479e09ef754.mp4

**Comment by kdev** [comment]
Nice work! ~~Is the bot open-source?~~ Found the repo.

**Reply by thboss** [reply]
The bot is based on csgo league bot https://github.com/csgo-league/csgo-league-bot
I will share the source code once it complete

**Comment by PhlexPlexico** [comment]
Wow that's awesome! By chance is it set to default 1v1s? Or is that configurable as well?

**Reply by thboss** [reply]
The default is 5v5, you can change the capacity,  and you can create multiple pugs at the same time, for example you can create 1v1, 2v2, and 5v5 pugs, also the bot is friendly with multiple discord servers 🙂

**Reply by thboss** [reply]
another thing, you can register different user api/key for each discord server

**Reply by PhlexPlexico** [reply]
Oh that's awesome! Glad to see it's mutli-server friendly as well, I was planning to work on that for my bot, but this looks awesome! Can't wait to see the source when you're done as well :)

**Reply by PhlexPlexico** [reply]
Also just curious, does it set the `is_pug` flag to 1 when creating a match? If so, it should delete the teams after a match is finished, so it avoids clogging up the teams table!

**Reply by thboss** [reply]
I almost done!, just wanna do some tests but your web looks broken  https://phlex.avidgamers.me 🤔

**Reply by thboss** [reply]
> Oh that's awesome! Glad to see it's mutli-server friendly as well, I was planning to work on that for my bot, but this looks awesome! Can't wait to see the source when you're done as well :)

Yeah it's even possible to make a public bot since anyone can register their user api data when there's public web as well :)

**Reply by thboss** [reply]
> Also just curious, does it set the `is_pug` flag to 1 when creating a match? If so, it should delete the teams after a match is finished, so it avoids clogging up the teams table!

Yes it did but I think it wasn't deleting teams after the match is finished (I'll confirm when the web back)

**Reply by PhlexPlexico** [reply]
Yup! I think I tested this the other day here as well:

https://github.com/PhlexPlexico/G5API/blob/f139f8ac230799d6326dabdf071a9a0b3aaeb21b/routes/legacy/api.js#L201-L214

And yeah, I think my friend who is generously hosting my API has been having some power issues for the past few days, so I think we may just have to wait until it gets sorted 😅

**Reply by PhlexPlexico** [reply]
Okay, the API should be up again!

**Reply by thboss** [reply]
Great! I just shared the source code https://github.com/thboss/CSGO-PUGs-Bot/tree/G5
only one thing, I get match config error in `/api/matches/{match_id}/config` when I remove this line https://github.com/thboss/CSGO-PUGs-Bot/blob/7087d6debb635ceb27ad52815e9b64f0cb58cceb/bot/cogs/utils/api.py#L116

**Reply by PhlexPlexico** [reply]
AH! Yeah, that's because flags are required in order to properly submit to the database. Given that a flag is needed, you could maybe just make a default flag for wherever the region that the bot is hosted? I don't think it really matters, so you could just default it to CA (Canada), US, or GE even, and that should be fine. 

https://github.com/PhlexPlexico/G5API/blob/f139f8ac230799d6326dabdf071a9a0b3aaeb21b/routes/teams.js#L358-L381

That's why, there is no "default" for a flag, so I would say maybe allow the user to set the team flag default?

**Reply by thboss** [reply]
I could add user flag to the users table https://github.com/thboss/CSGO-PUGs-Bot/blob/0729b920231c8e366bf55c781a450d4bb5ecd145/migrations/20200513_01_kPWNp-create-base-tables.py#L46
So when the user link their discord it's required to set their flag, and then the team flag is the captain's flag

**Reply by PhlexPlexico** [reply]
That would be perfect! I think I may swap out my bot that I'm using here for yours as well, but I also have a question - does this allow a user in discord to link their steam account with their discord? I am going to dig into your code here to check but that was the one thing I needed for PUGs was a users Steam ID and the only way I could really get around it was with a link system, like [so](https://github.com/PhlexPlexico/G5-Bot/blob/master/cogs/linksystem.py) stored in a SQLite database.

**Reply by thboss** [reply]
Of course, there is link system, https://github.com/thboss/CSGO-PUGs-Bot/blob/0729b920231c8e366bf55c781a450d4bb5ecd145/bot/cogs/commands.py#L157

So the user should enter a valid steam id/profile to link their discord account, and the bot built on postgreSQL as well

**Reply by PhlexPlexico** [reply]
I'll have to see about getting a postgresql server setup then so I can host my own instance, thanks for this!!

**Comment by PhlexPlexico** [comment]
Hey @thboss quick question in regards to setup. Is the `G5API_URL= # URL where the web panel is hosted` option set for the actual API endpoint, in this case `https://phlex.avidgamers.me/api` or just `https://phlex.avidgamers.me`?

Thanks!

**Reply by thboss** [reply]
it should be `https://phlex.avidgamers.me`
I just added validation for the flag in the last commit, and I highly recommend you update to the latest version, and make sure you use the G5 branch as well https://github.com/thboss/CSGO-PUGs-Bot/tree/G5

**Reply by PhlexPlexico** [reply]
Sounds good. And all these config settings, are they surrounded by quotes? I.e.
```
G5API_URL=https://phlex.avidgamers.me
```
Or
```
G5API_URL="https://phlex.avidgamers.me"
```

**Reply by thboss** [reply]
Both could work

**Reply by thboss** [reply]
I've caught a bug, I've set `min_player_ready` but its value didn't change in the match config https://phlex.avidgamers.me/api/matches/184
it should be 2 instead of 5 
```py
data = {
    'user_id': auth['user_id'],
    'user_api': auth['api_key'],
    'server_id': server[0]['id'],
    'team1_id': team1_id,
    'team2_id': team2_id,
    'title': '[PUG] Map {MAPNUMBER} of {MAXMAPS}',
    'is_pug': 1,
    'start_time': datetime.datetime.utcnow().strftime("%Y-%m-%d %H:%M:%S"),
    'ignore_server': 0,
    'max_maps': 1,
    'veto_mappool': map_pick.dev_name,
    'skip_veto': 1,
    'side_type': 'always_knife',
    'players_per_team': len(team_one),
    'min_player_ready': len(team_one + team_two)
}
```

**Reply by thboss** [reply]
Also, the `end_time` still `null` after teams forfeit!

**Reply by PhlexPlexico** [reply]
How are you forfeiting the match? As for the min_player, I think it's a different call, let me check here.

**Reply by thboss** [reply]
Players didn't ready up in time before the warm-up ends

**Reply by PhlexPlexico** [reply]
Hmm, by chance do you know what webcall or event get5 emits when that happens?

**Reply by thboss** [reply]
It was `end_series` event, and the get5 match status has changed to none.

**Reply by PhlexPlexico** [reply]
Can you do `get5_debug 16` and see if it makes an API call? I'm not sure if the timing out calls a forfeit or not. Also, `min_player_ready` should only be set to 5 max, since that's number of team members on each team to ready.

**Reply by PhlexPlexico** [reply]
Also the `min_player_ready` issue, I think the call from the POST requires `min_players_to_ready`, notice the extra S in players, that should fix that bug.

**Reply by PhlexPlexico** [reply]
Ah, I found a bug in the plugin here. It looks like The plugin actually doesn't call that post if the match gets cancelled... I think that should indeed be changed in favour of letting the API know the match was cancelled.

**Reply by PhlexPlexico** [reply]
Okay, so I think I understand the problem that's going on. Right now, if a user calls the cancel match command from G5V, it calls an end match command on the server, then goes and updates the data before. Calling this will create an error on the game server stating that the match is already finalized.  I remember running into this bug and now there's no easy way around it. I need to think of a way to wait for the game server to get a response before we forfeit or cancel a game on the `/matches/cancel` end point.

**Reply by PhlexPlexico** [reply]
[get5_apistats.zip](https://github.com/PhlexPlexico/G5API/files/6113102/get5_apistats.zip)

Here's an update web stats. I've also updated the API to not throw error on finalized matches when calling finish, but report that we have finished with a 200 and return.

**Reply by PhlexPlexico** [reply]
There! I believe that should work as intended now, thanks for pointing that out! :)

**Comment by thboss** [comment]
Great! the match now is canceling when not all players ready up in time.

**Comment by thboss** [comment]
And yeah `min_players_to_ready` works fine, the reason to set this value is we can pick up 1v1, 2v2... games through the bot (not only 5v5)

**Reply by PhlexPlexico** [reply]
Perfect, glad to see everything seems to be operational then! I'm going to try and setup my own instance here and have a community try it out. Would you mind cleaning up the excess teams that you have right now? If not, I can remove all the test cases if you'd like!

Thanks!

**Reply by thboss** [reply]
Ah of course! I forgot to remove them, sorry for that.

**Reply by PhlexPlexico** [reply]
No worries at all, thanks! 👍

**Comment by thboss** [comment]
I have a question, is that possible to add `rank` to the leaderboard?
That would be useful to implement auto balance for teams in the bot

**Reply by PhlexPlexico** [reply]
I would say maybe use the rating of an individual to determine their "rank" maybe? I was kind of against adding in anything that would differentiate from basic stats, since entering a whole rank system would be difficult to introduce, especially with weighting and all. Could you possibly just use ratings + maps played and generate a rank locally in the postgresql database?

**Reply by thboss** [reply]
Yeah that's possible, I could simply just use `avarage_rating` from `/api/leaderboard/players/pug` in order to rank players.

**Reply by thboss** [reply]
Hey @PhlexPlexico quick question about the leaderboard, how to get the number of maps that player played from the pugs?

**Reply by PhlexPlexico** [reply]
Great question. There actually is no report that returns the amount of maps played, just the amount of maps won... I suppose I could add that into the pig leaderboard call if you would like? It would just be a simple count of how many times they appear in the `player_stat` table where the `match_id` relates to a pug

**Reply by thboss** [reply]
That would be perfect!

**Reply by PhlexPlexico** [reply]
There, it's in there now as `totalMaps` :)

**Reply by thboss** [reply]
I don't see `totalMaps` here https://phlex.avidgamers.me/api/leaderboard/players/pug
Does your web up to date?
However, I added auto balance teams based on `average_rating`
https://github.com/thboss/CSGO-PUGs-Bot

**Reply by PhlexPlexico** [reply]
It should be up now! It's `total_maps`

**Reply by thboss** [reply]
Sounds like it works well for all players `/api/leaderboard/players` but for pugs `/api/leaderboard/players/pug` there are incorrect values, there are players have  `wins: 2` and  `total_maps: 1`

**Reply by PhlexPlexico** [reply]
Because wins is still counting as a global value both in and out of pugs... I could adjust it here as well!

E: It's now updated :)

**Comment by thboss** [comment]
I just added players stats commands :)

**Comment by kdev** [comment]
@thboss Good Job on that Discord Bot! Unfortunatly it is not possible to create any issues whatsoever to share some ideas or even raise an issue/bug?!

e.g. I would really like to see a Docker Image to easily deploy the Discord Bot to a personal host

**Comment by thboss** [comment]
Bot released here https://github.com/thboss/g5-discord-bot

**Comment by admrl** [comment]
Hi Everyone, is there a possibility that we create Discord roles and make them VIP perks just like in FACEIT? for example instances:

If player bought our VIP, they have perks like: Always Priority to be captain

but the question is what if all 10 players with VIP joined the queue, how do we randomize the VIP pick?

**Reply by PhlexPlexico** [reply]
If you're making feature requests, I would recommend creating an issue in the [proper](https://github.com/thboss/g5-discord-bot/issues) repository :)
